### ISSUE TITLE
subsystem-util: tests doesn't seem to run properly

### ISSUE THREAD

**Issue by ordian** [paritytech/polkadot#1584]
While trying to add a test for #1583, I've noticed that the tests also passes on master.
Digging a bit, it seems that in the select! https://github.com/paritytech/polkadot/blob/ae69d7a0dfe7c34b600df23fbfe3f5e0104d84ba/node/subsystem-util/src/lib.rs#L1049-L1055
`subsystem` finishes first even before the `test_task` creates a new job.

**Comment by coriolinus** [comment]
I wonder if it's helpful to replace the whole test harness with this one? https://github.com/paritytech/polkadot/blob/ddbf460664e1b76ea841f0937b0bffb9956de2e2/node/subsystem-test-helpers/src/lib.rs#L251-L281

Means integrating #1557 (or waiting for it to be merged), but might be worth it.

**Comment by ordian** [comment]
Unfortunately, it doesn't seem to be fixed by #1557. I'm looking into this.

**Comment by ordian** [comment]
Ok, so we're returning early here
https://github.com/paritytech/polkadot/blob/c39fbd38b5286b72cddbf8d582f12c3ac3c180aa/node/subsystem-util/src/lib.rs#L731

https://github.com/paritytech/polkadot/blob/c39fbd38b5286b72cddbf8d582f12c3ac3c180aa/node/subsystem-util/src/lib.rs#L634-L650

`jobs.outgoing_msgs` is an empty `StreamUnordered`. Citing from `streamunordered` docs:

> //! When a `StreamUnordered` is first created, it does not contain any streams. Calling `poll` in
//! this state will result in `Poll::Ready((None)` to be returned. Streams are submitted to the
//! set using `push`; however, the stream will **not** be polled at this point. `StreamUnordered`
//! will only poll managed streams when `StreamUnordered::poll` is called. As such, it is important
//! to call `poll` after pushing new streams.

@coriolinus any hints on how to fix this?

**Comment by coriolinus** [comment]
I think the fundamental issue is that we're `select`ing over the overseer, the test job, and the timeout:

https://github.com/paritytech/polkadot/blob/38024e28f353465a86bd36d151eb444d6a8fb2df/node/subsystem-test-helpers/src/lib.rs#L274-L280

I'm really confused at this point, because I could swear that I'd already implemented this fix, but what I think we want instead looks more like this:

```rust
futures::executor::block_on(async move { 
	if let Either::Right(_) = futures::future::select(
		futures::future::join(overseer, test),
		timeout,
	) { panic!("test timed out instead of completing") }
}); 
```

I'm freehanding, haven't run this through the compiler, but I think you get the idea. We don't want a early return from the overseer; we just want the first of either both the overseer and the test completing, or a timeout.

**Comment by ordian** [comment]
Sorry, I was pointing out to the `subsystem-util` code, not `subsystem-test-helpers`, where we're `select`ing over `subsystem`, the test job and the timeout. But the issue it seems is that the `Subsystem::run` can exit early because of the comment above.

I'm not sure how to is `subsystem-test-helpers` code different though, if we supply `Subsystem::run` instead of `overseer` it should be the same.

**Comment by coriolinus** [comment]
Right: the original problem is in the `subsystem-util::tests::test_harness` function. I suggested reimplementing that in terms of the `subsystem-test-helpers::subsystem-test-harness` function, but only just now realized that that function doesn't have the same join/select behavior that I thought it did.

There's nothing stopping you from reimplementing the `subsystem-util::tests::test_harness` function in the same way, though. The goal is just to ensure that the timeout can cancel everything, but that an early return from the overseer portion of the test harness doesn't cause an early return of the entire test.
